test/common/stats/thread_local_store_test.cc

```diff
@@ -460,6 +463,113 @@ TEST_F(StatsThreadLocalStoreTest, HotRestartTruncation) {
+      "noop");
+  store_->setStatsMatcher(std::make_unique<StatsMatcherImpl>(stats_config_));
+
+  // Testing No-op counters, gauges, histograms which match the prefix "noop".
+
+  // Counter
+  Counter& noop_counter = store_->counter("noop_counter");
+  EXPECT_EQ(noop_counter.name(), "");
+  EXPECT_EQ(noop_counter.value(), 0);
+  noop_counter.add(1);
+  EXPECT_EQ(noop_counter.value(), 0);
+  noop_counter.inc();
+  EXPECT_EQ(noop_counter.value(), 0);
+  noop_counter.reset();
+  EXPECT_EQ(noop_counter.value(), 0);
+  Counter& noop_counter_2 = store_->counter("noop_counter_2");
+  EXPECT_EQ(&noop_counter, &noop_counter_2);
+
+  // Gauge
+  Gauge& noop_gauge = store_->gauge("noop_gauge");
+  EXPECT_EQ(noop_gauge.name(), "");
+  EXPECT_EQ(noop_gauge.value(), 0);
+  noop_gauge.add(1);
+  EXPECT_EQ(noop_gauge.value(), 0);
+  noop_gauge.inc();
+  EXPECT_EQ(noop_gauge.value(), 0);
+  noop_gauge.dec();
+  EXPECT_EQ(noop_gauge.value(), 0);
+  noop_gauge.set(2);
+  EXPECT_EQ(noop_gauge.value(), 0);
+  noop_gauge.sub(2);
+  EXPECT_EQ(noop_gauge.value(), 0);
+  Gauge& noop_gauge_2 = store_->gauge("noop_gauge_2");
+  EXPECT_EQ(&noop_gauge, &noop_gauge_2);
+
+  // Histogram
+  Histogram& noop_histogram = store_->histogram("noop_histogram");
+  EXPECT_EQ(noop_histogram.name(), "");
+  Histogram& noop_histogram_2 = store_->histogram("noop_histogram_2");
+  EXPECT_EQ(&noop_histogram, &noop_histogram_2);
+
+  // Includes overflow stat.
+  EXPECT_CALL(*alloc_, free(_)).Times(1);
+
+  store_->shutdownThreading();
+}
+
+// We only test the exclusion list -- the inclusion list is the inverse, and both are tested in
+// test/common/stats:stats_matcher_test.
+TEST_F(StatsMatcherTLSTest, TestExclusionRegex) {
+  InSequence s;
+
+  EXPECT_CALL(*alloc_, alloc(_)).Times(2);
+
+  // Will block all stats containing any capital alphanumeric letter.
+  stats_config_.mutable_stats_matcher()->mutable_exclusion_list()->add_patterns()->set_regex(
+      ".*[A-Z].*");
+  store_->setStatsMatcher(std::make_unique<StatsMatcherImpl>(stats_config_));
+
+  // Testing No-op counters, gauges, histograms which match the prefix "noop".
+  Counter& lowercase_counter = store_->counter("lowercase_counter");
+  EXPECT_EQ(lowercase_counter.name(), "lowercase_counter");
+
+  Counter& uppercase_counter = store_->counter("UPPERCASE_counter");
+  EXPECT_EQ(uppercase_counter.name(), "");
+  uppercase_counter.inc();
+  EXPECT_EQ(uppercase_counter.value(), 0);
+  uppercase_counter.inc();
+  EXPECT_EQ(uppercase_counter.value(), 0);
+
+  // Adding another exclusion rule -- now we reject not just uppercase stats but those starting with
+  // the string "invalid".
+  stats_config_.mutable_stats_matcher()->mutable_exclusion_list()->add_patterns()->set_prefix(
+      "invalid");
+  store_->setStatsMatcher(std::make_unique<StatsMatcherImpl>(stats_config_));
+
+  Gauge& valid_gauge = store_->gauge("valid_gauge");
+  valid_gauge.set(2);
+  EXPECT_EQ(valid_gauge.value(), 2);
```

Comment:
AFAICT you don't have a similar valid test for regex counters or histograms.

Reply:
Good catch, I'll add single-rule tests for gauges + histograms and multiple-rule tests for counters + histograms.

Reply:
See a2f2f09.